### PR TITLE
Fix mix playlists

### LINKED ISSUE #555: Metadata can't be resolved for mix playlists
### Version

v6.0.2

### Details

When trying to get videos from a playlist I get a PlaylistUnavailableException in `public async ValueTask<PlaylistExtractor> GetPlaylistAsync(...)` [ YoutubeController.cs]
Examples:
https://www.youtube.com/watch?v=kKDO8Rxw3vs&list=RDkKDO8Rxw3vs&start_radio=1 -> Playlist 'RDkKDO8Rxw3vs' is not available.
https://www.youtube.com/watch?v=3U4lQ8xX3ag&list=RDEMPeuXuiddECzwmszjRj1EtQ  -> Playlist 'RDEMPeuXuiddECzwmszjRj1EtQ' is not available.

### Steps to reproduce

Insert one of the above mentioned URLs in Tyrrz/YoutubeDownloader.

### PR DESCRIPTION
**Pull request by xBaank**
Closes #555

Fix issue where metadata couldn't be resolved for mix playlists.

- Added ```https://www.youtube.com/youtubei/v1/next``` which seems to be the only endpoint to resolve mix playlits, as it can resolve any type of playlist  I changed the way playlistExtractor and playlistClient works. 
- Replaced  continuation token for visitorData (this is because continuationToken can't be used with Mix playlists).
- Added support to get thumbnails with system playlists (System playlists couldn't resolve thumbnails metadata). In the case of mix playlists it uses the first video thumbnail as youtube does too.
- Playlists metadata is tried to be resolved with the browse endpoint, if it can't resolve it (case of mix playlists) it will try to resolve it with next endpoint.
- Playlists url is now passed as a parameter to the constructor only in case that the browse endpoint can't get the playlists and gets it from next endpoint.

### CODE REVIEW

File: YoutubeExplode.Tests/TestData/PlaylistIds.cs [hunk 1]
```diff
@@ -11,4 +11,5 @@ internal static class PlaylistIds
     public const string UserUploads = "UUTMt7iMWa7jy0fNXIktwyLA";
     public const string Weird = "PL601B2E69B03FAB9D";
     public const string ContainsLongVideos = "PLkk2FsMngwGi9FNkWIoNZlfqglcldj_Zs";
+    public const string MixPlaylist = "RDTsYhxMnGYCw";
```
Comment (Tyrrrz) [inline YoutubeExplode.Tests/TestData/PlaylistIds.cs:14]: Can you please rename this to `VideoMix` and move it above `MusicMix`? Just to keep things consistent. All playlists starting with `RD` are mix playlists, but they behave a bit differently.

File: YoutubeExplode.DemoConsole/Program.cs [hunk 2]
```diff
@@ -21,6 +21,7 @@ public static async Task<int> Main()
         Console.Write("Enter YouTube video ID or URL: ");
         var videoId = VideoId.Parse(Console.ReadLine() ?? "");
 
+
```
Comment (Tyrrrz) [inline YoutubeExplode.DemoConsole/Program.cs:24]:
```suggestion

```

Remove extra linebreak

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 3]
```diff
@@ -20,6 +20,20 @@ internal partial class PlaylistExtractor
...
+            .GetPropertyOrNull("twoColumnWatchNextResults")?
+            .GetPropertyOrNull("playlist")?
+            .GetPropertyOrNull("playlist") ??
+
+        _content
+            .GetPropertyOrNull("continuationContents")?
+            .GetPropertyOrNull("playlistPanelContinuation")
+
+
+    );
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:35]:
```suggestion
            .GetPropertyOrNull("playlistPanelContinuation")
    );
```

Remove extra linebreak

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 4]
```diff
@@ -33,17 +47,19 @@ internal partial class PlaylistExtractor
             .ElementAtOrNull(1)?
             .GetPropertyOrNull("playlistSidebarSecondaryInfoRenderer")
     );
-
-    public bool IsPlaylistAvailable() => Memo.Cache(this, () =>
+    public bool IsPlaylistVideosAvailable() => Memo.Cache(this, () =>
+        TryGetPlaylistProperty() is not null
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:51]:
```suggestion

    public bool IsPlaylistVideosAvailable() => Memo.Cache(this, () =>
        TryGetPlaylistProperty() is not null
```

Add linebreak

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 5]
```diff
@@ -33,17 +47,19 @@ internal partial class PlaylistExtractor
...
+    );
+    public bool IsPlaylistDetailsAvailable() => Memo.Cache(this, () =>
         TryGetSidebar() is not null
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:54]:
```suggestion

    public bool IsPlaylistDetailsAvailable() => Memo.Cache(this, () =>
        TryGetSidebar() is not null
```

Add linebreak

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 6]
```diff
@@ -33,17 +47,19 @@ internal partial class PlaylistExtractor
...
 
     public string? TryGetPlaylistTitle() => Memo.Cache(this, () =>
-        TryGetSidebarPrimary()?
+        // next endpoint part
+        TryGetPlaylistProperty()?
             .GetPropertyOrNull("title")?
-            .GetPropertyOrNull("simpleText")?
             .GetStringOrNull() ??
-
+        // browse endpoint part
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:62]: Should we still have the `title.simpleText` as fallback?

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 7]
```diff
@@ -57,9 +73,17 @@ internal partial class PlaylistExtractor
...
+
     );
 
     public string? TryGetPlaylistAuthor() => Memo.Cache(this, () =>
+        //next enpoint part
+        TryGetPlaylistProperty()?
+            .GetPropertyOrNull("ownerName")?
+            .GetPropertyOrNull("simpleText")?
+            .GetStringOrNull() ??
+        //browse endpoint part
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:86]:
```suggestion
            .GetStringOrNull() ??

        //browse endpoint part
```

Linebreak

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 8]
```diff
@@ -106,60 +131,80 @@ internal partial class PlaylistExtractor
...
+                .GetPropertyOrNull("thumbnails")?
+                .EnumerateArrayOrNull()?
+                .Select(j => new ThumbnailExtractor(j))
+                .ToArray() ??
+
+        //next endpoint part
+        GetVideos()
+            .FirstOrDefault()?
+            .GetVideoThumbnails()
+            .ToArray() ??
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:148]: Should the `next` part be first, consistent with the others?

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 9]
```diff
@@ -106,60 +131,80 @@ internal partial class PlaylistExtractor
...
-        // Continuation response
-        _content
-            .GetPropertyOrNull("onResponseReceivedActions")?
-            .EnumerateArrayOrNull()?
-            .ElementAtOrNull(0)?
-            .GetPropertyOrNull("appendContinuationItemsAction")?
-            .GetPropertyOrNull("continuationItems")
+    public string? TryGetPlayListUrl() => Memo.Cache(this, () =>
+     TryGetPlaylistProperty()?
+         .GetPropertyOrNull("playlistShareUrl")?
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:155]: Does this work for both old and new approaches?

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 10]
```diff
@@ -106,60 +131,80 @@ internal partial class PlaylistExtractor
...
+         .GetStringOrNull()
     );
 
     public IReadOnlyList<PlaylistVideoExtractor> GetVideos() => Memo.Cache(this, () =>
-        TryGetContentRoot()?
+        TryGetPlaylistProperty()?
+            .GetPropertyOrNull("contents")?
             .EnumerateArrayOrNull()?
-            .Select(j => j.GetPropertyOrNull("playlistVideoRenderer"))
+            .Select(j => j.GetPropertyOrNull("playlistPanelVideoRenderer"))
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:163]: Does this work for both old and new approaches?

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 11]
```diff
@@ -106,60 +131,80 @@ internal partial class PlaylistExtractor
...
 
+    public string? TryGetLastVideoId() => Memo.Cache(this, () =>
+        TryGetPlaylistProperty()?
+            .GetPropertyOrNull("contents")?
+            .EnumerateArrayOrNull()?
+            .Select(j => j.GetPropertyOrNull("playlistPanelVideoRenderer"))
+            .LastOrDefault()?
+            .GetPropertyOrNull("videoId")?
+            .GetStringOrNull()
+    );
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:179]: Can we just use `GetVideos()` here?

File: YoutubeExplode/Channels/ChannelClient.cs [hunk 12]
```diff
@@ -83,6 +83,6 @@ private Channel Extract(ChannelPageExtractor channelPage)
     {
         // Replace 'UC' in channel ID with 'UU'
         var playlistId = "UU" + channelId.Value.Substring(2);
-        return new PlaylistClient(_http).GetVideosAsync(playlistId, cancellationToken);
+        return new PlaylistClient(_http).GetVideosAsync(playlistId, cancellationToken : cancellationToken);
```
Comment (Tyrrrz) [inline YoutubeExplode/Channels/ChannelClient.cs:86]: Was this change necessary?

```suggestion
        return new PlaylistClient(_http).GetVideosAsync(playlistId, cancellationToken);
```

File: YoutubeExplode/Playlists/Playlist.cs [hunk 13]
```diff
@@ -13,7 +13,7 @@ public class Playlist : IPlaylist
     public PlaylistId Id { get; }
 
     /// <inheritdoc />
-    public string Url => $"https://www.youtube.com/playlist?list={Id}";
+    public string Url => _url ?? $"https://www.youtube.com/playlist?list={Id}";
```
Comment (Tyrrrz) [inline YoutubeExplode/Playlists/Playlist.cs:16]: Would `https://www.youtube.com/watch?list=RDTsYhxMnGYCw` work? To avoid having to pass first video ID.

File: YoutubeExplode/Playlists/PlaylistClient.cs [hunk 14]
```diff
@@ -42,9 +43,10 @@ public class PlaylistClient
         var author = channelId is not null && channelTitle is not null
             ? new Author(channelId, channelTitle)
             : null;
+        //Can't get description from mix playlists
+        var description = playlistExtractor.TryGetPlaylistDescription();
```
Comment (Tyrrrz) [inline YoutubeExplode/Playlists/PlaylistClient.cs:47]: Defaulting to empty string is fine here, we were already doing it before in similar cases

File: YoutubeExplode/Playlists/PlaylistClient.cs [hunk 15]
```diff
@@ -67,24 +69,29 @@ public class PlaylistClient
...
     }
 
     /// <summary>
     /// Enumerates batches of videos included in the specified playlist.
     /// </summary>
     public async IAsyncEnumerable<Batch<PlaylistVideo>> GetVideoBatchesAsync(
         PlaylistId playlistId,
-        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+        [EnumeratorCancellation] CancellationToken cancellationToken = default       
+        )
```
Comment (Tyrrrz) [inline YoutubeExplode/Playlists/PlaylistClient.cs:83]:
```suggestion
        [EnumeratorCancellation] CancellationToken cancellationToken = default)
```

File: YoutubeExplode/Playlists/PlaylistClient.cs [hunk 16]
```diff
@@ -143,12 +150,20 @@ public class PlaylistClient
                 );
 
                 videos.Add(video);
+
+                
             }
 
+            lastVideoId = playlistExtractor.TryGetLastVideoId();
+            index = playlistExtractor.TryGetLastIndex() ?? 0;
+            visitorData  = visitorData ?? playlistExtractor.TryGetVisitorData();
```
Comment (Tyrrrz) [inline YoutubeExplode/Playlists/PlaylistClient.cs:159]: What does `visitorData` actually look like?

File: YoutubeExplode/Playlists/PlaylistController.cs [hunk 17]
```diff
@@ -47,16 +47,65 @@ public PlaylistController(HttpClient http)
...
+        if (!playlist.IsPlaylistDetailsAvailable())
+        {
+            //try getting info from the next endpoint if this one can't get the playlist
+            playlist = await GetPlaylistVideosAsync(playlistId,cancellationToken);
+        }
+
+        return playlist;
+    }
+
+    public async ValueTask<PlaylistExtractor> GetPlaylistVideosAsync(
```
Comment (Tyrrrz) [inline YoutubeExplode/Playlists/PlaylistController.cs:59]: Can we use this for all playlists, including for getting metadata?

**Review by Tyrrrz — COMMENTED** [review]
Thank you, this is valuable! Left some comments.

File: YoutubeExplode/Playlists/Playlist.cs (hunk 13 above)
Comment (xBaank) [reply YoutubeExplode/Playlists/Playlist.cs:16]: It does, but the first video shows as unavailable on the current device. So i think is better to pass the first video ID in the url in this case.

File: YoutubeExplode/Playlists/PlaylistClient.cs (hunk 16 above)
Comment (xBaank) [reply YoutubeExplode/Playlists/PlaylistClient.cs:159]: ```"visitorData": "CgtOM25XYWtXNmJBTSis3ISPBg%3D%3D"```
Seems to be like a token, we need to use it  along with the index and lastVideoId if we want to get the next videos on the playlists.

File: YoutubeExplode/Playlists/PlaylistController.cs (hunk 17 above)
Comment (xBaank) [reply YoutubeExplode/Playlists/PlaylistController.cs:59]: I thought we could, but the problem I faced was that it didn't return the playlist description.

File: YoutubeExplode/Bridge/PlaylistExtractor.cs (hunk 9 above)
Comment (xBaank) [reply YoutubeExplode/Bridge/PlaylistExtractor.cs:156]: playlistShareUrl always has this format ```https://www.youtube.com/watch?v=ID&list=ID```, this method does only work with next endpoint responses and next endpoint is used as a fallback of browse enpoint when trying to fetch metadata for mixes playlists.

File: YoutubeExplode/Playlists/PlaylistClient.cs (hunk 16 above)
Comment (xBaank) [reply YoutubeExplode/Playlists/PlaylistClient.cs:158]: And visitorData only comes in the first response,  if we didn't use it responses would have all the time the same videos even if we changed the index and lastVideoId (at least for mix playlists)).

File: YoutubeExplode/Bridge/PlaylistExtractor.cs (hunk 9 above)
Comment (Tyrrrz) [reply YoutubeExplode/Bridge/PlaylistExtractor.cs:159]: Do we actually need to extract this considering we already know the first video ID? We can just create the URL ourselves and not rely on `platlistShareUrl` being in the payload.

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 18]
```diff
@@ -106,60 +141,70 @@ internal partial class PlaylistExtractor
...
+        GetVideos()?
+            .LastOrDefault()?
+            .TryGetVideoId()
+    );
+
+    public int? TryGetLastIndex() => Memo.Cache(this, () =>
+        TryGetPlaylistProperty()?
+            .GetPropertyOrNull("contents")?
+            .EnumerateArrayOrNull()?
+            .Select(j => j.GetPropertyOrNull("playlistPanelVideoRenderer"))
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:185]: This could probably be replaced with `GetVideos` too?

File: YoutubeExplode/Bridge/PlaylistExtractor.cs [hunk 19]
```diff
@@ -106,60 +141,70 @@ internal partial class PlaylistExtractor
...
+            .Select(j => j.GetPropertyOrNull("nextContinuationData"))
             .WhereNotNull()
-            .FirstOrNull()?
-            .GetPropertyOrNull("continuationEndpoint")?
-            .GetPropertyOrNull("continuationCommand")?
-            .GetPropertyOrNull("token")?
+            .FirstOrDefault()
+            .GetPropertyOrNull("continuation")?
+            .GetStringOrNull()
+    );
```
Comment (Tyrrrz) [inline YoutubeExplode/Bridge/PlaylistExtractor.cs:202]: Are we actually passing this anywhere now?

File: YoutubeExplode/Playlists/Playlist.cs [hunk 20]
```diff
@@ -37,13 +39,15 @@ public class Playlist : IPlaylist
         string title,
         Author? author,
         string description,
-        IReadOnlyList<Thumbnail> thumbnails)
+        IReadOnlyList<Thumbnail> thumbnails,
+        string? url = null)
```
Comment (Tyrrrz) [inline YoutubeExplode/Playlists/Playlist.cs:43]: We'd have to pass `url` in a separate overload (with and without `url`) due to backwards compatibility

File: YoutubeExplode/Playlists/PlaylistController.cs [hunk 21]
```diff
@@ -47,16 +47,65 @@ public PlaylistController(HttpClient http)
         var raw = await SendHttpRequestAsync(request, cancellationToken);
         var playlist = PlaylistExtractor.Create(raw);
 
-        if (!playlist.IsPlaylistAvailable())
+        if (!playlist.IsPlaylistDetailsAvailable())
+        {
+            //try getting info from the next endpoint if this one can't get the playlist
+            playlist = await GetPlaylistVideosAsync(playlistId,cancellationToken);
+        }
```
Comment (Tyrrrz) [inline YoutubeExplode/Playlists/PlaylistController.cs:54]: Does this ever happen?

Comment (xBaank) [reply YoutubeExplode/Playlists/PlaylistController.cs:54]: yes, the playlist id that is used in #555 or this playlists id ```RDMM3Kuw0ZwI0WY``` would get the metadata from next endpoint.

File: YoutubeExplode/Playlists/Playlist.cs (hunk 20 above)
Comment (xBaank) [reply YoutubeExplode/Playlists/Playlist.cs:43]: do you mean removing it as an optional parameter and making another constructor overload?

File: YoutubeExplode/Bridge/PlaylistExtractor.cs (hunk 19 above)
Comment (xBaank) [reply YoutubeExplode/Bridge/PlaylistExtractor.cs:202]: Not really, so i could remove this method  as we don't use it anymore

File: YoutubeExplode/Bridge/PlaylistExtractor.cs (hunk 18 above)
Comment (xBaank) [reply YoutubeExplode/Bridge/PlaylistExtractor.cs:185]: Yes, just have to add a method that to ```PlaylistVideoExtractor``` that gets the video index.

File: YoutubeExplode/Playlists/Playlist.cs (hunk 20 above)
Comment (Tyrrrz) [reply YoutubeExplode/Playlists/Playlist.cs:43]: > do you mean removing it as an optional parameter and making another constructor overload?

Yeah. That's because other libraries that depend on YoutubeExplode will fail to bind to this constructor since the signature has changed.

File: YoutubeExplode/Bridge/PlaylistExtractor.cs (hunk 19 above)
Comment (Tyrrrz) [reply YoutubeExplode/Bridge/PlaylistExtractor.cs:202]: Yes, please remove any code that is not longer used in this new flow

File: YoutubeExplode/Bridge/PlaylistExtractor.cs (hunk 18 above)
Comment (Tyrrrz) [reply YoutubeExplode/Bridge/PlaylistExtractor.cs:185]: Yeah, let's do that
